src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp

```diff
@@ -9,56 +9,79 @@
+        nullptr,  // default security attributes
+        TRUE,     // manual reset event
+        FALSE,    // not set
+        nullptr); // name
+
+    // Use of TerminateThread for the file watcher thread was eliminated in favor of an event-based
+    // approach. Out of an abundance of caution, we are temporarily adding an environment variable
+    // to allow falling back to TerminateThread usage. If all goes well, this will be removed in a
+    // future release.
+    m_fRudeThreadTermination = false;
+    auto enableThreadTerminationValue = Environment::GetEnvironmentVariableValue(L"ASPNETCORE_FILE_WATCHER_THREAD_TERMINATION");
+    if (enableThreadTerminationValue.has_value())
+    {
+        m_fRudeThreadTermination = (enableThreadTerminationValue.value() == L"1");
+    }
 }
 
 FILE_WATCHER::~FILE_WATCHER()
 {
     StopMonitor();
-    WaitForMonitor(20); // wait for 1 second total
+    WaitForWatcherThreadExit();
 }
 
-void FILE_WATCHER::WaitForMonitor(DWORD dwRetryCounter)
+void FILE_WATCHER::WaitForWatcherThreadExit()
 {
-    if (m_hChangeNotificationThread != NULL)
+    if (m_hChangeNotificationThread == nullptr)
     {
-        DWORD dwExitCode = STILL_ACTIVE;
+        return;
+    }
 
-        while (!m_fThreadExit && dwRetryCounter > 0)
+    if (m_fRudeThreadTermination)
+    {
+        // This is the old behavior, which is now opt-in using an environment variable. Wait for
+        // the thread to exit, but if it doesn't exit soon enough, terminate it.
+        const int totalWaitTimeMs = 10000;
+        const int waitIntervalMs = 50;
+        const int iterations = totalWaitTimeMs / waitIntervalMs;
+        for (int i = 0; i < iterations && !m_fThreadExit; i++)
         {
-            if (GetExitCodeThread(m_hChangeNotificationThread, &dwExitCode))
-            {
-                if (dwExitCode == STILL_ACTIVE)
-                {
-                    // the file watcher thread will set m_fThreadExit before exit
-                    WaitForSingleObject(m_hChangeNotificationThread, 50);
-                }
-            }
-            else
+            // Check if the thread has exited.
+            DWORD result = WaitForSingleObject(m_hChangeNotificationThread, waitIntervalMs);
+            if (result == WAIT_OBJECT_0)
             {
-                // fail to get thread status
-                // call terminitethread
-                TerminateThread(m_hChangeNotificationThread, 1);
-                m_fThreadExit = TRUE;
+                // The thread has exited.
+                m_fThreadExit = true;
+                break;
             }
-            dwRetryCounter--;
         }
 
         if (!m_fThreadExit)
         {
+            LOG_INFO(L"File watcher thread did not exit. Forcing termination.");
             TerminateThread(m_hChangeNotificationThread, 1);
         }
     }
+    else
+    {
+        // Wait for the thread to exit.
+        WaitForSingleObject(m_hChangeNotificationThread, INFINITE);
```

Comment:
```suggestion
        LOG_INFO(L"Waiting for File watcher thread to exit.");
        WaitForSingleObject(m_hChangeNotificationThread, INFINITE);
```
So we know if it's hanging. The thread itself logs when it exits.